src/main/java/fi/aalto/cs/apluscourses/intellij/model/IntelliJModelFactory.java

```diff
@@ -22,11 +28,43 @@ public IntelliJModelFactory(@NotNull Project project) {
   public Course createCourse(@NotNull String name,
                              @NotNull List<Module> modules,
                              @NotNull Map<String, String> requiredPlugins) {
-    return new IntelliJCourse(name, modules, requiredPlugins, project);
+    IntelliJCourse course = new IntelliJCourse(name, modules, requiredPlugins, project);
+    // Add a module change listener with the created course instance to the project
+    project.getMessageBus().connect().subscribe(ProjectTopics.MODULES, new ModuleListener() {
+      @Override
+      public void moduleRemoved(@NotNull Project project,
+                                @NotNull com.intellij.openapi.module.Module projectModule) {
+        Module module = course.getModuleOpt(projectModule.getName());
+        if (module != null) {
+          course.onModuleRemove(module);
+        }
+      }
+    });
+    project.getMessageBus().connect().subscribe(VirtualFileManager.VFS_CHANGES,
+        new BulkFileListener() {
+          @Override
+          public void after(@NotNull List<? extends VFileEvent> events) {
+            for (VFileEvent event : events) {
+              if (event instanceof VFileDeleteEvent) {
+                String deletedFileName = event.getFile().getName();
+                Module module = course.getModuleOpt(deletedFileName);
+                if (module != null) {
+                  course.onModuleFilesDeletion(module);
+                }
+              }
+            }
+          }
+        }
+    );
```

Comment:
I'm re-reading the use-cases of the [`Course.getModuleOpt(...)`](https://github.com/Aalto-LeTech/intellij-plugin/blob/c42d7a5b2fdd0b5c5c70e6ac8e88905bd902be46/src/main/java/fi/aalto/cs/apluscourses/model/Course.java#L201-L209). 

In case of [`moduleRemoved(...)`](https://github.com/Aalto-LeTech/intellij-plugin/blob/c42d7a5b2fdd0b5c5c70e6ac8e88905bd902be46/src/main/java/fi/aalto/cs/apluscourses/intellij/model/IntelliJModelFactory.java#L37) it seems OK, we take a projectModule's name and compare it (inside `getModuleOpt` with names of existing modules)

However, in case of [`after()`](https://github.com/Aalto-LeTech/intellij-plugin/blob/1691959d898f8d1d612cb132773aae36d0366e59/src/main/java/fi/aalto/cs/apluscourses/intellij/model/IntelliJModelFactory.java#L50) I'm a bit puzzled. So overall, for the cases, where there is a bulk change of files occur, we check all the events (is there by accident a narrower _topic_ to listen?) and then pass it down to the `getModuleOpt`. So far OK, but then, inside the `getModuleOpt(...)` the **deletedFileName** (name of the file affected) being [compared to the **name of the module**](https://github.com/Aalto-LeTech/intellij-plugin/blob/1691959d898f8d1d612cb132773aae36d0366e59/src/main/java/fi/aalto/cs/apluscourses/model/Course.java#L206), so I wonder if it will ever work?

Could there be a test for such a use case or something?

Reply:
Ideally we would only listen to file deletion events here since we only care about those, but I haven't found a way to do that. However, I'll double-check. 

Comparing the deleted file name to the name of the module works, because in our case the directory of the module always has the same name as the module. For an example, going to the project directory and deleting the folder `GoodStuff` causes a file deletion event with the filename `"GoodStuff"` (same as the name of the module). 

However, deleting files from the `GoodStuff` directory won't work. For an example, deleting `GoodStuff.iml` doesn't affect the modules in anyway, because no module has the name `"GoodStuff.iml"`. This could cause problems, but I'm not sure if we need to deal with cases where students delete individual files from modules.